### PR TITLE
[flutter_web_plugins] Migrate to null safety.

### LINKED ISSUE #69843: Migrate flutter_web_plugins package to null safety
This is a requirement for https://github.com/flutter/flutter/issues/66395

## Proposal

Migrate the `flutter_web_plugins` package to null safety.

### PR DESCRIPTION
**Pull request by ditman**
## Description

This PR migrates the `flutter_web_plugins` to null safety, and solves issues reported by the analyzer.

The functionality stays the same. Null behavior should also stay the same as before; new implementation should be fully backwards-compatible.

## Related Issues

* Fixes https://github.com/flutter/flutter/issues/69843

## Tests

Tests are also updated, and they all pass:

```
$ flutter test --platform chrome

Building flutter tool...
Running "flutter pub get" in flutter_web_plugins...                395ms
Throwing away cached asset graph because the language version of some package(s) changed.
This would most commonly happen when updating dependencies or changing your min sdk constraint.

00:16 +20: All tests passed!                    
```

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_web_plugins/pubspec.yaml
Comment (jonahwilliams) [inline packages/flutter_web_plugins/pubspec.yaml:7]: If you change the constraints to https://github.com/flutter/flutter/blob/master/packages/flutter/pubspec.yaml#L8 you should not need the experiment (though you may need to restart the IDE)

File: packages/flutter_web_plugins/lib/src/plugin_registry.dart
Comment (jonahwilliams) [inline packages/flutter_web_plugins/lib/src/plugin_registry.dart:11]: nit: bug or username for the todo

File: packages/flutter_web_plugins/lib/src/plugin_event_channel.dart
Comment (jonahwilliams) [inline packages/flutter_web_plugins/lib/src/plugin_event_channel.dart:108]: I think we've been leaving the asserts in with an ignore, because users that are not opted into null safety can still pass null in

File: packages/flutter_web_plugins/analysis_options.yaml
Comment (jonahwilliams) [inline packages/flutter_web_plugins/analysis_options.yaml:1]: See comment below on sdk constraint, this should not be necessary anymore

**Review by blasten — APPROVED** [review]
LGTM after @jonahwilliams comments are addressed

File: packages/flutter_web_plugins/pubspec.yaml
Comment (ditman) [reply packages/flutter_web_plugins/pubspec.yaml:7]: Ahhh I remember trying `2.12.0` and having the compiler complain about my version of dart not being recent enough (but I didn't do the `-0`!) 

I'll try that again.

Comment (ditman) [reply packages/flutter_web_plugins/pubspec.yaml:7]: This *does work*, but it's uncovered a warning that I hadn't seen before:

```
The package resolution is not fully migrated to null-safety.            
                                                                        
package:test_api is not opted into null safety in its pubspec.yaml:     
  |                                                                     
8 |   sdk: '>=2.10.0-0 <2.12.0'                                         
  |        ^^^^^^^^^^^^^^^^^^^^                                         
  |                                                                     
                                                                        
Either downgrade your sdk constraint, or invoke dart/flutter with       
`--no-sound-null-safety`.                                               
                                                                        
To learn more about available versions of your dependencies try running 
`pub outdated --mode=null-safety`.                                      
                                                                        
See more at https://dart.dev/null-safety/migration-guide.               
```

Comment (ditman) [reply packages/flutter_web_plugins/pubspec.yaml:7]: I have this (autogenerated) dependency on test_api:

```
  test_api: 0.2.19-nullsafety.4 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
```

File: packages/flutter_web_plugins/lib/src/plugin_registry.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/plugin_registry.dart:11]: Yes, I'll blame the line

File: packages/flutter_web_plugins/lib/src/plugin_event_channel.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/plugin_event_channel.dart:108]: Ah, very good point, I'll get those back in!

File: packages/flutter_web_plugins/pubspec.yaml
Comment (jonahwilliams) [reply packages/flutter_web_plugins/pubspec.yaml:7]: It is a bug in pub, I believe it can be ignored

Comment (ditman) [reply packages/flutter_web_plugins/pubspec.yaml:7]: > **0.2.19-nullsafety.5**: Update SDK constraints to `>=2.12.0-0 <3.0.0` based on beta release guidelines.

Comment (ditman) [reply packages/flutter_web_plugins/pubspec.yaml:7]: I tried to update-packages, but test_api is stuck in .4?

<details>

<summary><tt>$ flutter update-packages --force-upgrade</tt></summary>

```
Building flutter tool...
Upgrading packages...
WARNING: the following packages use hard-coded version constraints:
  - flutter_gallery_assets: ^0.2.0
  - mockito: 4.1.1
  - vm_service_client: 0.2.6+2
  - flutter_template_images: 1.0.1
  - shelf: 0.7.5
  - async: 2.5.0-nullsafety.2
Skipping boolean_selector because it was not transitive
  - characters: 1.1.0-nullsafety.4
Skipping charcode because it was not transitive
  - clock: 1.1.0-nullsafety.2
  - collection: 1.15.0-nullsafety.4
  - fake_async: 1.2.0-nullsafety.2
  - js: 0.6.3-nullsafety.2
Skipping matcher because it was not transitive
  - meta: 1.3.0-nullsafety.5
  - path: 1.8.0-nullsafety.2
Skipping pedantic because it was not transitive
  - pool: 1.5.0-nullsafety.2
Skipping source_maps because it was not transitive
Skipping source_map_stack_trace because it was not transitive
Skipping source_span because it was not transitive
  - stack_trace: 1.10.0-nullsafety.5
  - stream_channel: 2.1.0-nullsafety.2
  - string_scanner: 1.1.0-nullsafety.2
Skipping term_glyph because it was not transitive
  - test: 1.16.0-nullsafety.7
  - test_api: 0.2.19-nullsafety.4
  - test_core: 0.3.12-nullsafety.7
  - typed_data: 1.3.0-nullsafety.4
  - vector_math: 2.1.0-nullsafety.4
  - platform: 3.0.0-nullsafety.3
  - file: 6.0.0-nullsafety.3
  - process: 4.0.0-nullsafety.3
  - process_runner: 4.0.0-nullsafety.4
  - build_runner_core: 5.2.0
  - build_modules: 2.10.1
  - path_provider: 1.6.14
  - video_player: 2.0.0-nullsafety.1
  - url_launcher: 6.0.0-nullsafety
  - connectivity: 3.0.0-nullsafety
  - device_info: 2.0.0-nullsafety
Running "flutter pub upgrade" in flutter_update_packages.CDJCJS...      2,540ms
Running "flutter pub get" in snippets...                           413ms
Running "flutter pub get" in manual_tests...                       313ms
Running "flutter pub get" in gen_keycodes...                       252ms
Running "flutter pub get" in vitool...                             364ms
Running "flutter pub get" in tools...                              518ms
Running "flutter pub get" in devicelab...                          493ms
Running "flutter pub get" in web...                                306ms
Running "flutter pub get" in platform_interaction...               495ms
Running "flutter pub get" in android_views...                    1,791ms
Running "flutter pub get" in flutter_driver_screenshot_test...      1,297ms
Running "flutter pub get" in android_semantics_testing...        1,291ms
Running "flutter pub get" in flutter_gallery...                  1,470ms
Running "flutter pub get" in external_ui...                      1,248ms
Running "flutter pub get" in gradle_deprecated_settings...         625ms
Running "flutter pub get" in channels...                         1,317ms
Running "flutter pub get" in flutterapp...                         650ms
Running "flutter pub get" in image_loading...                    1,242ms
Running "flutter pub get" in flavors...                          1,247ms
Running "flutter pub get" in release_smoke_test...                 910ms
Running "flutter pub get" in flutterapp...                         771ms
Running "flutter pub get" in web_compile_tests...                  589ms
Running "flutter pub get" in abstract_method_smoke_test...         526ms
Running "flutter pub get" in ios_platform_view_tests...          1,166ms
Running "flutter pub get" in hybrid_android_views...             1,522ms
Running "flutter pub get" in splash_screen_load_rotate...          902ms
Running "flutter pub get" in splash_screen_kitchen_sink...       1,328ms
Running "flutter pub get" in splash_screen_trans_rotate...       1,138ms
Running "flutter pub get" in ui...                               1,248ms
Running "flutter pub get" in ios_app_with_extensions...            669ms
Running "flutter pub get" in non_nullable...                       711ms
Running "flutter pub get" in web_e2e_tests...                    1,225ms
Running "flutter pub get" in android_embedding_v2_smoke_test...        762ms
Running "flutter pub get" in tracing_tests...                      716ms
Running "flutter pub get" in stocks...                           1,102ms
Running "flutter pub get" in microbenchmarks...                  1,259ms
Running "flutter pub get" in complex_layout...                   1,314ms
Running "flutter pub get" in macrobenchmarks...                  1,153ms
Running "flutter pub get" in platform_views_layout...            1,178ms
Running "flutter pub get" in platform_views_layout_hybrid_composition...      1,255ms
Running "flutter pub get" in customer_testing...                   705ms
Running "flutter pub get" in automated_tests...                  1,110ms
Running "flutter pub get" in bots...                             1,294ms
Running "flutter pub get" in platform_view...                      526ms
Running "flutter pub get" in hello_world...                      1,187ms
Running "flutter pub get" in platform_channel_swift...           1,441ms
Running "flutter pub get" in platform_channel...                 1,316ms
Running "flutter pub get" in flutter_view...                       492ms
Running "flutter pub get" in layers...                             735ms
Running "flutter pub get" in image_list...                         713ms
Running "flutter pub get" in splash...                             618ms
Running "flutter pub get" in flutter_driver...                   1,122ms
Running "flutter pub get" in fuchsia_remote_debug_protocol...      1,231ms
Running "flutter pub get" in main...                               272ms
Running "flutter pub get" in flutter_tools...                      895ms
Running "flutter pub get" in test...                               699ms
Running "flutter pub get" in test_private...                       507ms
Running "flutter pub get" in flutter...                            896ms
Running "flutter pub get" in _flutter_web_build_script...        1,251ms
Running "flutter pub get" in flutter_web_plugins...                391ms
Running "flutter pub get" in flutter_goldens...                    856ms
Running "flutter pub get" in flutter_localizations...              685ms
Running "flutter pub get" in project_root...                       164ms
Running "flutter pub get" in flutter_test...                       644ms
Running "flutter pub get" in flutter_goldens_client...             476ms
Downloading lcov data for package:flutter...                       508ms

Ran 'pub' 64 times and fetched coverage data in 57.7s.
```

</details>

Comment (ditman) [reply packages/flutter_web_plugins/pubspec.yaml:7]: My branch was created from an older version of flutter. I rebased the branch to the latest tip of master, and this went away.

**Comment by ditman** [comment]
For this to be finished, flutter deps need to be updated at least to:

```
  source_span: 1.8.0-nullsafety.4 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
```

Once that's in, this will be mergeable (Can the update be done in a separate PR?)

**Comment by jonahwilliams** [comment]
https://github.com/flutter/flutter/commit/e375651cc542edeebbd74d7c12a120d958928111 was reverted, so we need to figure out what the error is before this can land

**Comment by jonahwilliams** [comment]
doing another dependency roll today

File: packages/flutter_web_plugins/lib/src/navigation/js_url_strategy.dart
Comment (yjbanov) [inline packages/flutter_web_plugins/lib/src/navigation/js_url_strategy.dart:43]: Unfortunately, the initial version of nullability cannot express "if the argument is not null, the return result is also not null" so everyone needs to deal with the nullable return even when it's impossible. Would it be possible to make this function take and return non-null values and push the null awareness to the call sites?

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (yjbanov) [inline packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:15]: According to the dartdoc this function allows null argument.

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (yjbanov) [inline packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:16]: Nulls are allowed it seems, so `!` is probably incorrect.

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (yjbanov) [inline packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:317]: remove?

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (yjbanov) [inline packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:229]: Looks like `HashUrlStrategy` doesn't deal well with null `pathname` and `search`. Do we need to allow nulls here?

File: packages/flutter_web_plugins/lib/src/navigation/utils.dart
Comment (yjbanov) [inline packages/flutter_web_plugins/lib/src/navigation/utils.dart:7]: Statics are lazy-initialized by the language. We don't need to hand-code lazy initialization. We can do:

```
final AnchorElement _urlParsingNode = AnchorElement();
```

File: packages/flutter_web_plugins/lib/src/navigation/utils.dart
Comment (yjbanov) [inline packages/flutter_web_plugins/lib/src/navigation/utils.dart:17]: Making `_urlParsingNode` `final` non-null will remove the need for `!`s

Comment (yjbanov) [inline packages/flutter_web_plugins/lib/src/navigation/utils.dart:17]: Is `?? ''` the right fallback? In the previous version `pathname` continued to be null. This also ties into the previous question about whether `pathname` should ever be null.

File: packages/flutter_web_plugins/lib/src/navigation/utils.dart
Comment (yjbanov) [inline packages/flutter_web_plugins/lib/src/navigation/utils.dart:21]: I'd make this:

```dart
final Element? _baseElement = document.querySelector('base');
```

That would simplify a lot of code below

File: packages/flutter_web_plugins/lib/src/navigation/utils.dart
Comment (yjbanov) [inline packages/flutter_web_plugins/lib/src/navigation/utils.dart:33]: Making `_baseElement` `final` would simplify this method greatly:

```dart
return _baseElement?.getAttribute('href');
```

File: packages/flutter_web_plugins/lib/src/plugin_event_channel.dart
Comment (yjbanov) [inline packages/flutter_web_plugins/lib/src/plugin_event_channel.dart:136]: Is awaiting on null allowed? If so, then this entire `if` block can be just:

```dart
await subscription?.cancel();
```

Comment (ditman) [reply packages/flutter_web_plugins/lib/src/plugin_event_channel.dart:136]: Good point, I don't know!

File: packages/flutter_web_plugins/lib/src/navigation/js_url_strategy.dart
Comment (mdebbar) [reply packages/flutter_web_plugins/lib/src/navigation/js_url_strategy.dart:43]: That's a good idea. This function is only called in one place (and is not exposed outside of this package) so it should be straightforward to move the null check to the call site here: https://github.com/flutter/flutter/blob/ffcf41913655409e11c3cf2c28b08af906a5143e/packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart#L18

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (mdebbar) [reply packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:15]: Yes, null is a valid value here to disable url strategies altogether. The signature should be:
```dart
void setUrlStrategy(UrlStrategy? strategy) {
```

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (mdebbar) [reply packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:16]: Let's check for `strategy == null` here and avoid passing it through `convertToJsUrlStrategy` if it's null.

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (mdebbar) [reply packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:317]: Oops, this was my bad. Please remove.

**Comment by jonahwilliams** [comment]
Version updates have landed BTW

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:15]: Ah, makes sense, this also requires the js-interop typedef to be:

```
typedef _JsSetUrlStrategy = void Function(JsUrlStrategy?);
```

So I can pass 'null' to the JS.

File: packages/flutter_web_plugins/lib/src/navigation/js_url_strategy.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/js_url_strategy.dart:43]: I like these conversion functions not having to handle the null case; this is going to be a very common pattern in the web plugins, that have *plenty* of these!

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:16]: This is done too, thanks for noticing the `!`, I totally misread this API!

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:317]: Wow, good catch, I didn't even notice that comment!

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:229]: In this case, I tried to preserve the logic from dart:html; [`Location.search`][search] and [`Location.pathname`][pathname] are both defined as `String?`.

At first I thought it was because their behavior is not defined when they're missing, but pathname should return an empty string when undefined.

Anyway, I can modify this and make it non-null, and then ensure that we return the correct defaults in the specific implementations.

[search]: https://developer.mozilla.org/en-US/docs/Web/API/Location/search
[pathname]: https://developer.mozilla.org/en-US/docs/Web/API/Location/pathname

Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:229]: I made all these non-null in the abstract class, and provided sane defaults in the specific implementation. The rest of the class works better now, less `!` thrown around.

File: packages/flutter_web_plugins/lib/src/navigation/utils.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/utils.dart:17]: In this case, pathname comes from dart:html `class AnchorElement`, and it's [defined as `String?`][pathname]. In my tests it's never null, in fact, if no pathname is supplied, in chrome it returns `'/'`. According to the spec, it should return `''`.

Looking at the next line of code (which I haven't touched), it seems that `pathname` will *never* be null, regardless of what the definition says. I'd say an empty string is a nice fallback.

_(Just to double check, you can't call isEmpty on null)_
```
Uncaught TypeError: Cannot read property 'get$isEmpty' of nullError: TypeError: Cannot read property 'get$isEmpty' of null
```

[pathname]: https://api.dart.dev/stable/2.10.3/dart-html/AnchorElement/pathname.html

Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/utils.dart:17]: I had to eagerly initialize `_urlParsingNode` for this to work, but it's now final and non-null.

File: packages/flutter_web_plugins/lib/src/navigation/utils.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/utils.dart:7]: #TIL this is lazily-initialized!

File: packages/flutter_web_plugins/lib/src/navigation/utils.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/utils.dart:21]: Reading some more code below (`checkBaseHref`), it seems this `base` element needs to be pre-rendered on the index.html, so this should work as expected!

File: packages/flutter_web_plugins/lib/src/navigation/utils.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/utils.dart:33]: Yes, it becomes so simple that it can be defined with a fat arrow and still fit in the line width :) Done that.

File: packages/flutter_web_plugins/lib/src/plugin_event_channel.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/plugin_event_channel.dart:136]: It seems you *can* await on null (it'll resolve to null):

```dart
void main() async {
  Future<int>? bah = null;
  print(await bah); // prints null
  print('done'); // prints done
}
```

**Comment by ditman** [comment]
I think I addressed all of @yjbanov's comments above!

@jonahwilliams, thanks for updating the deps! I think this will still fail analysis, because it seems `flutter_test` is still not migrated (or it seemed so while I was testing this with the latest SDK constraints and without the analysis_options.yaml file)

**Comment by ditman** [comment]
I guess things might still break if changes happen in flutter_test or flutter, which are still not fully migrated to NNBD. (I hope they're easily fixable since everything lives in the same repo!)

**Comment by jonahwilliams** [comment]
flutter and flutter_test should be fully migrated to null safety

**Comment by ditman** [comment]
@jonahwilliams OK, then it's probable that their sdk constraint is not being detected as null-safe; I get this warning every time I `pub get`:

```
$ flutter pub get
Building flutter tool...
The package resolution is not fully migrated to null-safety.            
                                                                        
package:flutter_test is not opted into null safety in its pubspec.yaml: 
  |                                                                     
5 |   sdk: ">=2.10.0-0.0.dev <3.0.0"                                    
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^                                    
  |                                                                     
                                                                        
Either downgrade your sdk constraint, or invoke dart/flutter with       
`--no-sound-null-safety`.                                               
                                                                        
To learn more about available versions of your dependencies try running 
`pub outdated --mode=null-safety`.                                      
                                                                        
See more at https://dart.dev/null-safety/migration-guide.               
Running "flutter pub get" in flutter_web_plugins...              2,449ms
```

File: packages/flutter_web_plugins/pubspec.yaml
Comment (ditman) [reply packages/flutter_web_plugins/pubspec.yaml]: (This is done)

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (mdebbar) [inline packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:301]: FWIW, in the engine, we are adding a `!` to fix this:

https://github.com/flutter/engine/blob/acaef6f00d78da2f9a8f0fafb36d1c45e6b2513d/lib/web_ui/lib/src/engine/navigation/url_strategy.dart#L268

But I'll leave it up to you what you prefer to do here :)

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (ditman) [inline packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:271]: Checking if getBaseHref can be null

Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:271]: It can:

https://github.com/flutter/engine/blob/master/lib/web_ui/lib/src/engine/navigation/url_strategy.dart#L246

**Review by mdebbar — APPROVED** [review]
url strategy changes look good to me. Thanks for migrating!

File: packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart
Comment (ditman) [reply packages/flutter_web_plugins/lib/src/navigation/url_strategy.dart:301]: I prefer having a default value, since the dart:html API seems to suggest that there's the chance that `null` may be returned? (maybe they know about some browser quirk that we don't?)

**Comment by jonahwilliams** [comment]
You need to pull in the latest master

**Comment by ditman** [comment]
Rebased all the way to https://github.com/flutter/flutter/commit/a3f6ea6f38df520915566eca5cda16827477a251.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by kevmoo** [comment]
Woo hoo!

When can we expect a dev release cut w/ this commit?

**Comment by ditman** [comment]
@kevmoo I'm not sure about the roll to dev, maybe "Flutter Rollers" knows?

**Pull request merged**
